compiler/src/iree/compiler/Dialect/Stream/Transforms/UnifyEncodingForGlobals.cpp

```diff
@@ -0,0 +1,330 @@
+              auto globalOp =
+                  symbolTable.lookup<IREE::Util::GlobalOpInterface>(globalName);
+              if (!globalOp) {
+                LDBG() << "      Bail: global not found: " << globalName;
+                return false;
+              }
+              if (globalOp.isGlobalMutable()) {
+                LDBG() << "      Bail: mutable source global";
+                return false;
+              }
+              if (foundSourceGlobal && foundSourceGlobal != globalOp) {
+                LDBG() << "      Bail: multiple source globals";
+                return false;
+              }
+              foundSourceGlobal = globalOp;
+              auto initAttr = globalOp.getGlobalInitialValue();
+              if (initAttr) {
+                // Has inline initial value, done tracing.
+                return true;
+              }
+              // No inline initial value; trace through stores in initializers.
+              IREE::Util::GlobalTable globalTable(moduleOp);
+              globalTable.rebuild();
+              auto &global = globalTable.lookup(globalName);
+              for (auto storeOp : global.storeOps) {
+                worklist.push_back(storeOp.getStoredGlobalValue());
+              }
+              return true;
+            })
+            .Case([&](IREE::Stream::TensorConstantOp) {
+              // Constant is a valid leaf source.
+              return true;
+            })
+            .Default([&](Operation *op) {
+              LDBG() << "      Bail: unknown op " << op->getName();
+              return false;
+            });
+    if (!shouldContinue) {
+      return nullptr;
+    }
+  }
+  return foundSourceGlobal;
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+struct UnifyEncodingForGlobalsPass
+    : public impl::UnifyEncodingForGlobalsPassBase<
+          UnifyEncodingForGlobalsPass> {
+  void runOnOperation() override {
+    ModuleOp moduleOp = getOperation();
+    GlobalEncodingAnalyzer analyzer(moduleOp);
+    if (failed(analyzer.run())) {
+      LDBG() << "Analysis failed, skipping.";
+      return;
+    }
+    auto candidates = analyzer.getSourcesWithMultipleEncodings();
+    if (candidates.empty()) {
+      LDBG() << "No source globals with multiple encodings found.";
+      return;
+    }
+    LDBG() << "Found " << candidates.size()
+           << " source globals with multiple encodings:";
+    for (auto name : candidates) {
+      LDBG() << "  - " << name;
+    }
+
+    // Unify encodings for each source global with multiple encodings.
+    for (auto sourceName : candidates) {
+      const auto &sourceInfo =
+          analyzer.getSourceGlobals().find(sourceName)->second;
+
+      // TODO(#22485): Select unified encoding via resolver. For now, use
+      // identity.
+      auto unifiedEncoding =
```

Comment:
So this is just for initial commit right? Cause I dont know what "identity encoding" means.

Reply:
Yes, this is just for initial commit. In some follow-ups, we'll let encoding resolver to provide the preferred encoding, using new interface methods. They are not developed yet, so we start with identity encoding. A tensor type with identity encoding will be resolved into the same tensor without encodings. I.e., I imagine that they will be resolved to no-encodings if the interface methods are not implemented yet. This is just a first step for testing; we'll improve this in follow-ups.